### PR TITLE
Speed up multi part parsing #1027

### PR DESCRIPTION
**Pull request by LukeButtersFunnelback**
(See issue https://github.com/eclipse/jetty.project/issues/1027)
In essence we speed up multipart requests by reducing the amount of branching done. This is done by reading chunks (byte arrays) of data at once rather than a single byte at a time. 

Although this does result in faster multi part parsing (1.3s to 0.3s for about 3k requests) it does lead to some sticky situations where we need to undo how much we have read. To do this I have used mark/reset however for base64 encoding this is not possible and instead careful work had to be done to ensure each read would return a very specific amount of data. Also in this implementation we can write one char to much which may result in a file being created to early (e.g. we are supposed to create a file at 50bytes but we create it at 49), as we write one extra char we have to do some special work to backup in the case we have written to a file we truncate the file otherwise we backup the count into the byte array output stream.

### CODE REVIEW

**Comment by joakime** [comment]
@LukeButtersFunnelback please sign the Eclipse CLA and submit the commits with `-s` (sign-off).

_(otherwise we cannot accept this PR due to Eclipse Legal reasons)_

**Comment by LukeButtersFunnelback** [comment]
I went to eclipse and accepted the CLA and then made a commit with -s
```
commit ec970785d6463ed689656c6ea64a4cae08c1b480
Author: Luke Butters <lbutters@funnelback.com>
Date:   Wed Oct 26 09:01:37 2016 +1100

    Fixed spelling and signed off
    
    Signed-off-by: Luke Butters <lbutters@funnelback.com>
```
Note sure why it is not accepting it the details have:
```
The following users have invalid Signed-off-by footers. Please use the Signed-Off-By (git commit -s) mechanism to confirm the origin of the submission. The identity in the Signed-off-By must match the Author or the Committer.
    lbutters@funnelback.com
```

**Comment by gregw** [comment]
Thanks for this contribution.   Note that in 9.4.x or 10.x we need to totally replace this parse with a fully asynchronous version as that will be needed for 4.0 Servlet API.

@janbartel can you review as well when you get a chance.

File: jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java [hunk 1]
```diff
@@ -665,117 +678,235 @@ public int read() throws IOException
                             }
                             return c;
                         }
+                        
+                        @Override
+                        public int read(byte b[]) throws IOException {
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: Can you comply with the coding standard in this file:  { on a new line

File: jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java [hunk 2]
```diff
@@ -642,11 +655,11 @@ else if(tl.startsWith("filename="))
                 InputStream partInput = null;
                 if ("base64".equalsIgnoreCase(contentTransferEncoding))
                 {
-                    partInput = new Base64InputStream((ReadLineInputStream)_in);
+                    partInput = new Base64InputStream(is);
                 }
                 else if ("quoted-printable".equalsIgnoreCase(contentTransferEncoding))
                 {
-                    partInput = new FilterInputStream(_in)
+                    partInput = new FilterInputStream(is)
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: Would it not be better to use the JVMs BufferedInputStream as the basis rather than doing your own buffering?  The buf filed is protected in that class and thus available to be used.

**Comment by LukeButtersFunnelback** [comment]
I have hopefully placed all the `{` on a new line, I don't know how to get the eclipse ip validation check to pass.

File: jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java (hunk 2 above)
Comment (LukeButtersFunnelback) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: I don't think I follow. I used mark when possible as `_in` is already a BufferedInputStream.

Comment (LukeButtersFunnelback) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: (I wonder if you didn't see my second comment from clicking start a review rather than Add single comment).
Also I suspect a bug exists here, I will write a test

Comment (LukeButtersFunnelback) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: Ignore that I wrote a test and it seems fine.

File: jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java [hunk 3]
```diff
@@ -665,117 +678,275 @@ public int read() throws IOException
...
-                        lf=(c==10||state==10);
-                        if(state==10)
-                            state=-2;
                     }
                 }
                 finally
                 {
                     part.close();
+                    //In the case of a file we may need to trim the last char.
+                    if(trimLastChar && part._file != null && part._file.exists())
```
Comment (LukeButtersFunnelback) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: What are your thoughts on doing this?

File: jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java [hunk 4]
```diff
@@ -665,117 +678,275 @@ public int read() throws IOException
...
+                    int offsetIntoBufin = -1;
+                    int bufInSize;
+                    while(true) 
                     {
-                        int b=0;
-                        while((c=(state!=-2)?state:partInput.read())!=-1)
+                        //If mark is not supported the stream must ensure that we will
+                        //never read too much, i.e. not more than the boundary (excluding the trailing new line
+                        // or double dash).
+                        if(partInput.markSupported())
```
Comment (LukeButtersFunnelback) [inline jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java]: Here is where we use mark if it is possible.

**Comment by gregw** [comment]
I checked lbutters@funnelback.com at https://projects.eclipse.org/user/cla/validate and  there is a valid CLA(ECA) registered.  The commits have the signed-off-by line, so that should be OK?     @joakime any ideas?

Luke,   I got interrupted in my review... I try again tomorrow to give a more sane response.

**Comment by LukeButtersFunnelback** [comment]
Ok cool, happy to make further changes as you see fit.

**Comment by gregw** [comment]
Luke, we've not looked at this for a bit as coincidentally we had some other multi-part issues to investigate/fix.

We are keen to improve performance, so if you do get a chance to merge to the latest that would be good.

**Comment by LukeButtersFunnelback** [comment]
I will resolve the conflicts soon, I need to finish up some other work first

**Comment by janbartel** [comment]
Luke, I'm finally able to tune in a bit to this pull request - thanks for looking into this.

As the conflicts prevent applying the pull request, even locally, I cloned your repo, checked out your branch and ran the MultipartInputStream tests as a starting point.

I got 1027 repetitions of this log line:
"WARN:oeju.MultiPartInputStreamParser:main: Badly formatted multipart request"

The current jetty-9.3 code produces 6 of these log lines.

There were also some differences in the memory utilization. I got this:
SUREFIRE-859: [GC (Allocation Failure) [PSYoungGen: 262144K->2924K(305664K)] 262144K->2932K(1005056K), 0.0042522 secs] [Times: user=0.02 sys=0.00, real=0.01 secs] 
SUREFIRE-859: [GC (Allocation Failure) [PSYoungGen: 265068K->1696K(305664K)] 265076K->1712K(1005056K), 0.0044592 secs] [Times: user=0.02 sys=0.00, real=0.01 secs]
and also
SUREFIRE-859:  PSYoungGen      total 305664K, used 166910K [0x00000000eab00000, 0x0000000100000000, 0x0000000100000000)

I think the difference here is in your new testVaryingChunkSize() test - I haven't had a chance to look further at that yet, but just wanted to flag it up so you can take a look as you work on the merge conflicts. 

I'll grab some more cycles and tune in to the proposed code changes.

**Comment by janbartel** [comment]
@LukeButtersFunnelback the multipart parsing is a bit impenetrable at the best of times .... could you summarize here where you think you are getting the performance improvement?  If the input stream is a BufferedInputStream, I can't see the benefit of doing your own additional buffering.  That leads me to think that perhaps an improvement you're seeing is not in the reading, and maybe its in writing more than one char at a time to the output stream?

**Comment by LukeButtersFunnelback** [comment]
Yes I think one if the existing tests already prints that log line, and I added another test which called that test many many times. Hopefully within the next few days (exclude weekend) I should be able to resolve the conflicts. I think however it is worth reviewing this code, perhaps a fundamental problem exists with this solution or you have some improvements to the code.

**Comment by LukeButtersFunnelback** [comment]
The benefit comes from reducing the number of branches on the inner most loops. From memory this is achieved by:
1. Removed a branch on each read [see](https://github.com/LukeButtersFunnelback/jetty.project/blob/a1a4d3f6d87553d58c1dcb164b5554f65a0339f2/jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java#L745) The branch is moved to outside of the loop.
2. Reduced the number branches on each read by only checking the against the maxRequestSize every 4096 bytes rather than every single char read [see](https://github.com/LukeButtersFunnelback/jetty.project/blob/a1a4d3f6d87553d58c1dcb164b5554f65a0339f2/jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java#L835)
3. Reduce the number of branches on write, as we no longer write one char at a time we no longer need to do [the branches here](https://github.com/LukeButtersFunnelback/jetty.project/blob/a1a4d3f6d87553d58c1dcb164b5554f65a0339f2/jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java#L133) for every char and instead we do them once per 4096 chars. This will also speed up actually writing to the underlying output stream as it can do less branching ensuring it has enough space.
4. When handling `\r\n` this code backups on the `OutputStream` or '`File` if it wrote out an extra '\r', by doing this we again reduce branches.

I think you state it correctly in that the real advantage is we buffer bytes before we write them out to the output stream. I guess yes we do re-buffer the input stream however we can not always assume we are reading a buffered input stream it might be a base64 decoding stream or a quoted-printable stream.

**Comment by janbartel** [comment]
@LukeButtersFunnelback can you also clarify about the early file creation: you said this impl might create a file earlier than it should. So eg if a file was supposed to be created after 50 bytes, it is created after 49 bytes instead. Is it possible that the 50th byte is never forthcoming and that an empty file is left behind?

**Comment by LukeButtersFunnelback** [comment]
@janbartel the problem occurs when a part ends with \r\n e.g:
```
Data here\n
\r\n--boundary
```
The bytes written to the output stream would be `Data here\n\r`, if we told the parser to write out a file when it had `11` bytes we would find that the bytes would be written to a file, however when the parser checks the last char and seeing that it is `\r` it would backup one char in the case of the file trimming the file by one char. Thus resulting in a file that is `10`bytes long. 
I could probably fix that up if you wish.

**Comment by LukeButtersFunnelback** [comment]
I tried to get the latest changes into my fork/branch but I screwed up. Would someone be kind enough to tell me what to do?

**Comment by janbartel** [comment]
@LukeButtersFunnelback  what seems to be the problem?

Assuming it is a problem with the git setup:

Have a look at [fork a repo](https://help.github.com/articles/fork-a-repo/), particularly steps 7 and 8, and then at [sync a fork](https://help.github.com/articles/syncing-a-fork/) to pull in latest changes to your fork.

If there's something more specific, let me know and I'll try to help.

**Comment by janbartel** [comment]
@LukeButtersFunnelback  re the file size, can you double check that you are not assuming that there will always be a /r/n combo? We also accept just /r and just /n and mixtures.  Look particularly at the tests testLFOnlyRequest(), testCROnlyRequest(), testCRandLFMixRequest().

**Comment by LukeButtersFunnelback** [comment]
Hi Jan, I tried to get the changes but my PULL request turned into something with 72 modified files :(
Also Jan this code does work wirth `\r\n`,`\r`, `\n` the tests seem to verify that

**Comment by gregw** [comment]
@janbartel what are your thoughts on this PR?   Are you concerned about the over eager consumption of bytes?   Do you think we should perhaps just defer until the total rewrite that is needed for async multi part support?

**Comment by janbartel** [comment]
I'm still evaluating it against what we already have. My main concern is that the parsing of the part is already a bit complex and the PR code is even more so, so more difficult to maintain. I know we have a number of users with very odd-ball multipart requests, so I want to make sure we understand exactly what this PR code does in case we come across a case that we haven't already covered in our unit tests.

I should have a fuller analysis by the end of week, although deferring this to a re-write to support async multipart is indeed tempting.

**Comment by LukeButtersFunnelback** [comment]
Happy to change the code to not create a file 1 byte to early.

**Comment by LukeButtersFunnelback** [comment]
For maintenance is it a problem for me to be included in coming up with the fix?

**Comment by janbartel** [comment]
@LukeButtersFunnelback there's no problem with you being involved, in fact, thanks for contributing to jetty.

**Comment by gregw** [comment]
@LukeButtersFunnelback if you like, rather than fix this old clunky parser, we could collaborate on creating an entirely new parser that would work asynchronously.

Essentially we need a new parser that would look substantially like HttpParser, but would emit events for Parts rather than requests and responses.    This could then be used for both the existing blocking and the new async Part APIs.

It should be a much faster approach as it would be able to avoid parsing line and then parsing parts.

Although if you are happy to fix @janbartel's concerns about the 1 byte on this PR, we can progress this instead... or as well.

**Comment by LukeButtersFunnelback** [comment]
How do I get involved? Has someone already written a prototype?

If I stopped this from creating a file one byte to early would that help get this merged? or is there a fundamental objection to the MR that everyone is too polite to mention.

**Comment by gregw** [comment]
@LukeButtersFunnelback I defer to @janbartel about if the one byte issue is enough to get this merged. The concern is that we don't want to change semantics.

To get involved in creating the async parser, well nobody has started work on that yet.    I have mocked up some APIs that we may need to implement for the Servlet 4.0 expert group, and while they have not yet been accepted, they've not been rejected either.  So if you are interested, I could create a branch off master and put in some of this skeleton API in there.  You could then fork that branch and work on a parser (I can point you at similar code to work from).   We could then comment/refine/collaborate on that branch.

This is code that we will eventually write ourselves, probably Q2 next year, but would definitely appreciate contributions to get it done earlier.

If you are doing lots of multi-part work, then having an async provider of Parts should also give you a good performance boost in your application.

**Comment by janbartel** [comment]
Hi @LukeButtersFunnelback , 

I'm going to close this pull request - time has moved on and we really need to concentrate on a fully asynchronous version instead. Be great if you had the time to contribute to that.

**Comment by LukeButtersFunnelback** [comment]
ok,

Just keep in mind a function call for every byte is not efficient, asking for a an array from a stream is likely much faster than asking for each byte.
